data_steward/analytics/cdr_ops/rdr_export_qc.py

```diff
@@ -0,0 +1,316 @@
+            dialect='standard')
+
+# # Check value_source_value unmapped to standard
+
+query = """
+SELECT 
+    SUM(CASE WHEN value_as_concept_id IS NULL THEN 1 ELSE 0 END) as n_null_concept_id
+    , SUM(CASE WHEN value_as_concept_id=0 THEN 1 ELSE 0 END) as n_zero_concept_id
+FROM `{dataset}.observation`
+WHERE value_source_value IS NOT NULL
+and value_source_value != ''
+"""
+pd.read_gbq(query.format(dataset=new_rdr),
+            project_id=project_id,
+            dialect='standard')
+
+# # check date = extract date datetime
+
+query = """
+SELECT 
+    SUM(CASE WHEN observation_date != EXTRACT(DATE FROM observation_datetime) THEN 1 ELSE 0 END) as n_date_datetime_issues
+FROM `{dataset}.observation`
+"""
+pd.read_gbq(query.format(dataset=new_rdr),
+            project_id=project_id,
+            dialect='standard')
+
+# # Check for duplicates
+
+query = f"""
+with duplicates AS (
+    SELECT
+        person_id
+        , observation_datetime
+        , observation_source_value
+        , value_source_value
+        , value_as_number
+        , value_as_string
+        --, questionnaire_response_id
+        , COUNT(1) as n_data
+    FROM `{new_rdr}.observation`
+    INNER JOIN `{cope_version_map}` USING(questionnaire_response_id) -- For COPE only
+    GROUP BY 1,2,3,4,5,6
+)
+SELECT 
+    n_data as duplicates
+    , COUNT(1) as n_duplicates
+FROM duplicates
+WHERE n_data > 1
+GROUP BY 1
+ORDER BY 2 DESC
+"""
+pd.read_gbq(query, project_id=project_id, dialect='standard')
+
+# # Check if numeric data in value_as_string
+
+query = f"""
+SELECT
+    observation_source_value
+    , COUNT(1) AS n
+FROM `{new_rdr}.observation`
+WHERE SAFE_CAST(value_as_string AS INT64) IS NOT NULL
+GROUP BY 1
+ORDER BY 2 DESC
+"""
+pd.read_gbq(query, project_id=project_id, dialect='standard')
+
+# # Check all questionnaire_id in map
+
+query = f"""
+SELECT
+    COUNT(1)
+FROM `{new_rdr}.observation`
+INNER JOIN `pipeline_tables.cope_concepts` on observation_source_value = concept_code
+WHERE questionnaire_response_id NOT IN (SELECT questionnaire_response_id FROM `{cope_version_map}`)
+"""
+pd.read_gbq(query, project_id=project_id, dialect='standard')
+
+# # Check questionnaire_response_ids don't overlap amongst versions
+
+query = f"""
+SELECT
+    participant_id,
+    questionnaire_response_id,
+    cope_month
+FROM `{cope_version_map}`
+"""
+version = pd.read_gbq(query, project_id=project_id, dialect='standard')
+
+v_dict = {}
+v_dict['may'] = set(version[version['cope_month'] == 'may']
+                    ['questionnaire_response_id'].unique())
+v_dict['june'] = set(version[version['cope_month'] == 'june']
+                     ['questionnaire_response_id'].unique())
+v_dict['july'] = set(version[version['cope_month'] == 'july']
+                     ['questionnaire_response_id'].unique())
+
+# +
+from itertools import combinations
+
+for d1, d2 in combinations(v_dict.keys(), 2):
+    print(f"{d1}-{d2}: {len(v_dict[d1].intersection(v_dict[d2]))}")
```

Comment:
```suggestion
query = f"""
SELECT
    questionnaire_response_id,
    COUNT(*) n
FROM `{cope_version_map}` 
GROUP BY questionnaire_response_id
HAVING n > 1
"""
pd.read_gbq(query, project_id=project_id, dialect='standard')
```
@rfrancis1 would it suffice to check for duplicate `questionnaire_response_id`? The only way there would be overlap between versions is if it repeated and as far as I know multiple `cope_version_map` rows should never have the same `questionnaire_response_id`.

Reply:
That might work, Mark.